### PR TITLE
Improve error handling.

### PR DESCRIPTION
**Pull request by jaydenseric**
* Refactor to use fewer Busboy event listeners.
* Improved error handling, fixing [#26](https://github.com/jaydenseric/apollo-upload-server/issues/26):
  * [Misordered multipart fields](https://github.com/jaydenseric/graphql-multipart-request-spec) cause `processRequest` to throw a middleware error.
  * The `map` field provided by the client is used to naively check the `maxFiles` option is not exceeded for an immediate middleware error, even though the real number of files to parse might not match in a malformed request.
  * A `scalar Upload` promise rejects with an error if the file was missing from the request.
  * Already if a file exceeds the `maxFileSize` option the file is truncated, the stream emits a `limit` event and `stream.truncated === true`. Now an `error` event is also emitted with a meaningful message.
  * Aborting requests from the client causes `scalar Upload` promises to reject with a meaningful error for file upload streams that have not yet been parsed. For streams being parsed an `error` event is emitted and `stream.truncated === true`. It is up to consumers to cleanup aborted streams in their resolvers.

Todo:

* [x] Be sure of the logic of the errors thrown when the files limit is discovered to be exceeded in a malformed request (more files than declared by the client in the `map` field).
* [x] Ensure exported Koa and Express middleware handle errors thrown by `processRequest` correctly. Server error responses should have correct HTTP codes.
* [ ] Document error handling in the readme.
* [x] Add tests; there are now a lot of edge cases to cover.

### CODE REVIEW

File: src/index.mjs [hunk 1]
```diff
@@ -80,5 +175,8 @@ export const apolloUploadExpress = options => (request, response, next) => {
       request.body = body
       next()
     })
-    .catch(next)
+    .catch(error => {
+      if (error.status && error.expose) response.status(error.status)
```
Comment (jaydenseric) [inline src/index.mjs]: @AdamYee does this look like a good idea? Setting the response status automatically for errors in the Express middleware. I'm assuming it is an Express convention for middleware not to send the error response, that it is up to the consumer to set up a catch-all error handling middleware in their app?

Comment (AdamYee) [reply src/index.mjs]: I think the simple `.catch(next)` is what you want. Inspecting the `error` object is usually the responsibility of the error handling middleware from my experience. They might choose to overwrite the status anyways, or not send an error at all.

I would take a look at multer's error handling API. https://github.com/expressjs/multer/blob/master/index.js#L40. Maybe there's value in providing a callback into your error handling in case the consumer wants to do something specific?

**Comment by jaydenseric** [comment]
Help with the last test ("Abort uploads") would be *amazing;* I have spent hours going around in circles. It looks like the final hurdles are working out why the `error` event on the file stream is causing `Uncaught Exception` when it is for sure being listened, and why AVA is not detecting an end to the test.
